### DISCUSSION TITLE
`account.access_token` & `account.refresh_token` is `undefined` when using Twitter Provider with `next-auth`?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [nextauthjs/next-auth#2966] [Help] [answered]
I am following [this tutorial](https://spacejelly.dev/posts/how-to-make-twitter-api-requests-with-nextauth-js-session-tokens/) & am unable to get the value of `access_token` & `refresh_token` from [next-auth](https://next-auth.js.org/). My callback function looks like:

```ts
callbacks: {
    async jwt({ token, account, user }) {
      if (account) {
        token[account.provider] = {
          accessToken: account.access_token,
          refreshToken: account.refresh_token,
        }
        console.log(token[account.provider])
      }

      return token
    },
  },
```

It currently logs `{ accessToken: undefined, refreshToken: undefined }` to the console.

This is how it's shown [in the docs when using jwt callbacks](https://next-auth.js.org/configuration/callbacks#jwt-callback) but it's not working for me.

I have made a complete reproduction on branch `next-auth-with-twitter-api-v2` → https://github.com/deadcoder0904/twitter-api-v2-3-legged-login-using-next-connect/tree/next-auth-with-twitter-api-v2

**Comment by deadcoder0904** [comment]
@balazsorban44 I have asked about [this on Reddit /r/nextjs](https://www.reddit.com/r/nextjs/comments/q983mi/accountaccess_token_accountrefresh_token_is/) as well & I don't think anyone is able to make it work even though it worked with v3. Maybe it's a bug.

Can you @ someone who is familiar with Twitter Oauth API with `next-auth` to confirm this or should I create an issue of it from the discussion?

**Reply by deadcoder0904** [reply]
@balazsorban44 It looks like this is clearly a bug. It works in v3 (see branch `next-auth-v3-with-twitter-api-v2`) → https://github.com/deadcoder0904/twitter-api-v2-3-legged-login-using-next-connect/tree/next-auth-v3-with-twitter-api-v2

**Comment by balazsorban44** [comment]
this isn't a bug. there is no access or refresh tokens in case of Twitter, please refer to our previous discussions. (I cannot find them to link to right now)

**Reply by deadcoder0904** [reply]
@balazsorban44 then how come it works on v3 of `next-auth`, that's what i'm asking? the tutorial i linked above uses exactly that to make request.

the branch i made now uses the same access token & refresh token that you say don't exist & makes request to search endpoint & fetches new tweets. it wouldn't be able to work without them so again how does it work?

can you confirm with someone that uses twitter api or wrote the code for twitter api v4 in next-auth just to make sure because as you said you haven't worked with twitter api & multiple people including me think it's a bug.

**Reply by balazsorban44** [reply]
I did the refactoring, no one to ask. nothing changed with the v1 implementation apart from renaming the access token to `oauth_token` and the refresh token to `oauth_token_secret`. please stop asking me to do any work, the source code is available publicly, have a look. you also don't need to tag me in every response, I get a notification anyway. thank you for understanding.

I have repeatedly explained the difference between OAuth1 (Twitter) and the other providers and linked you to the resources I know exist.

I don't have the capacity to solve your problem further, I'm sorry.

If you aren't ready to migrate to v4, I recommend you keep using v3, since you say it works for you.

**Reply by deadcoder0904** [reply]
cool, got it!

**Reply by deadcoder0904** [reply]
> renaming the access token to oauth_token and the refresh token to oauth_token_secret

all i needed to know :)

**Comment by deadcoder0904** [comment, marked as answer]
So the `access_token` got renamed to `oauth_token` & `refresh_token` got renamed to `oauth_token_secret`:

```ts
callbacks: {
    async jwt({ token, account, user }) {
      if (account) {
        token[account.provider] = {
          accessToken: account.oauth_token,
          refreshToken: account.oauth_token_secret,
        }
      }

      return token
    },
  },
```

**Reply by balazsorban44** [reply]
Or to be more precise, they have been stopped being renamed to access and refresh token respectively.
